addons/stock/models/stock_move_line.py

```diff
@@ -147,17 +148,39 @@ def _onchange_serial_number(self):
                         message = _('You cannot use the same serial number twice. Please correct the serial numbers encoded.')
                     elif not self.lot_id:
-                        counter = self.env['stock.production.lot'].search_count([
-                            ('company_id', '=', self.company_id.id),
-                            ('product_id', '=', self.product_id.id),
-                            ('name', '=', self.lot_name),
-                        ])
-                        if counter > 0:
-                            message = _('Existing Serial number (%s). Please correct the serial number encoded.') % self.lot_name
+                        lots = self.env['stock.production.lot'].search([('product_id', '=', self.product_id.id),
+                                                                        ('name', '=', self.lot_name),
+                                                                        ('company_id', '=', self.company_id.id)])
+                        quants = lots.quant_ids.filtered(lambda q: q.quantity != 0 and q.location_id.usage in ['customer', 'internal', 'transit'])
+                        if quants:
+                            message = _('Serial number (%s) already exists in location(s): %s. Please correct the serial number encoded.', self.lot_name, ', '.join(quants.location_id.mapped('display_name')))
                 elif self.lot_id:
                     counter = Counter([line.lot_id.id for line in move_lines_to_check])
                     if counter.get(self.lot_id.id) and counter[self.lot_id.id] > 1:
                         message = _('You cannot use the same serial number twice. Please correct the serial numbers encoded.')
+                    else:
+                        # check if in correct source location
+                        quants = self.env['stock.quant'].search([('product_id', '=', self.product_id.id),
+                                                                 ('lot_id', '=', self.lot_id.id),
+                                                                 ('quantity', '!=', 0),
+                                                                 '|', ('location_id.usage', '=', 'customer'),
+                                                                      '&', ('company_id', '=', self.company_id.id),
+                                                                           ('location_id.usage', 'in', ('internal', 'transit'))])
+                        sn_locations = quants.mapped('location_id')
+                        if sn_locations and self.location_id not in sn_locations:
+                            location_within_picking_location = self.env['stock.location']
+                            if self.picking_id:
+                                for location in sn_locations:
+                                    if location == self.picking_id.location_id or location in self.picking_id.location_id.child_ids:
```

Comment:
Same comment about parent_path than stock_scrap.py

Reply:
Double dang, missed that one. Fixed it.